pkg/controller/jobframework/setup.go

```diff
@@ -99,6 +100,52 @@ func (m *integrationManager) setupControllers(mgr ctrl.Manager, log logr.Logger,
 	})
 }
 
+func (m *integrationManager) setupControllerAndWebhook(mgr ctrl.Manager, name string, fwkNamePrefix string, cb IntegrationCallbacks, options Options, opts ...Option) error {
+	if err := cb.NewReconciler(
+		mgr.GetClient(),
+		mgr.GetEventRecorderFor(fmt.Sprintf("%s-%s-controller", name, options.ManagerName)),
+		opts...,
+	).SetupWithManager(mgr); err != nil {
+		return fmt.Errorf("%s: %w", fwkNamePrefix, err)
+	}
+	if err := cb.SetupWebhook(mgr, opts...); err != nil {
+		return fmt.Errorf("%s: unable to create webhook: %w", fwkNamePrefix, err)
+	}
+	m.enableIntegration(name)
+	return nil
+}
+
+func waitForAPI(ctx context.Context, mgr ctrl.Manager, log logr.Logger, gvk schema.GroupVersionKind, retryInterval time.Duration, action func()) {
+	ticker := time.NewTicker(retryInterval)
+	defer ticker.Stop()
+	for {
+		_, err := getRESTMapping(mgr, gvk)
+		if err == nil {
+			action()
+			return
+		}
+		if !meta.IsNoMatchError(err) {
+			log.Error(err, "Failed to get REST mapping for gvk", "gvk", gvk)
+		}
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			continue
+		}
+	}
+}
+
+func getRESTMapping(mgr ctrl.Manager, gvk schema.GroupVersionKind) (*meta.RESTMapping, error) {
+	restMapperMutex.RLock()
```

Comment:
Would the `get` fail occasionally if there is no lock? Note that there is no such locking in production code, so such errors (race conditions) could happen in prod code. Seeing the errors which may be throws is desired, we can retry them in the test code, as the prod code would retry.

Reply:
The `get` never failed without the lock. In fact, all tests passed without the `-race` flag. 

This lock is added for the tests to pass with the `-race` flag. In the tests, a [`write` is performed](https://github.com/kubernetes-sigs/kueue/pull/2574/files#diff-0e0e2357162d6c7d6172e1f1e2f0fed3d27946005ee221718652d972f6a61879R185), causing a race condition here.

Reply:
I see, in that case we could retry the write, but then there is a question if the write would be retried by the production code. I think, ideally we would have an integration test for that. At the unit test level I prefer to retry the write.